### ISSUE TITLE
CtScanner which provides CtRole of scanned element

### ISSUE THREAD

**Issue by pvojtechovsky** [INRIA/spoon#1614]
It would be nice if CtScanner (or similar class) would report the role of the scanned element.
The code would simply look like this:
```java
class CtScanner {
	public void scan(Object o, CtRole role) {
		...
	}
...
	public <T> void visitCtClass(final CtClass<T> ctClass) {
		enter(ctClass);
		scan(ctClass.getAnnotations(), CtRole.ANNOTATION);
		scan(ctClass.getSuperclass(), CtRole.SUPER_TYPE);
		scan(ctClass.getSuperInterfaces(), CtRole.INTERFACE);
		scan(ctClass.getFormalCtTypeParameters(), CtRole.TYPE_PARAMETER);
		scan(ctClass.getTypeMembers(), CtRole.TYPE_MEMBER);
		scan(ctClass.getComments(), CtRole.COMMENT);
		exit(ctClass);
	}
...
}
```

I am opening this issue NOW, because it might influence the design of CtRole enum values.

Note: `scan(ctClass.getTypeMembers(), CtRole.TYPE_MEMBER);`

Here is again visible the conflict in current CtRole concept. In this case the roles should be defined for root attribute (TYPE_MEMBER) and not for all leafs (FIELD, CONSTRUCTOR, NESTED_TYPE, METHOD, ANNONYMOUS_EXECUTABLE).

So what about to have roles for all cases?
* TYPE_MEMBER - is general for FIELD, NESTED_TYPE, EXECUTABLE
* EXECUTABLE - is general for CONSTRUCTOR, METHOD, ANNONYMOUS_EXECUTABLE
* FIELD
* CONSTRUCTOR
* NESTED_TYPE
* METHOD
*  ANNONYMOUS_EXECUTABLE

**Comment by tdurieux** [comment]
I just thought something related, for example, we could also use the roles to refine the element that your want to scan.

Example:

```
CtStatement with CtRole.STATEMENT
CtExpression with CtRole.ASSIGNEMENT
...
```

**Comment by pvojtechovsky** [comment]
This issue is solved by #1632. May be just create new issue for the last comment of @tdurieux, before closing this one. I do not understand this request, so I do know if it is already solved or not.

**Comment by surli** [comment]
@tdurieux I close this one, reopen a new issue if necessary for your last comment: I confess I don't understand here either.
